File: src/lib/container.ts
// Composição das dependências (o papel do contexto do Spring). Criado sob demanda e uma única
// vez por processo; `globalThis` evita recriar pools a cada hot-reload no `next dev`.
import { randomUUID } from 'node:crypto';
import { loadConfig, type AppConfig } from './config';
import { createDatabase } from './db';
import type { Database } from './db-types';
import { RedisLockManager } from './lock';
import { createSimulatedGateway } from './payment-gateway';
import { createRedis, createRedisLockStore, type RedisClient } from './redis';
import { createPostgresRepositories } from './repositories';
import { createCatalogService, type CatalogService } from './services/catalog-service';
import { createPaymentService, type PaymentService } from './services/payment-service';
import { createReservationService, type ReservationService } from './services/reservation-service';
import { generateReservationToken } from './token';

export interface Container {
  config: AppConfig;
  db: Database;
  redis: RedisClient;
  catalog: CatalogService;
  reservations: ReservationService;
  payments: PaymentService;
}

function build(): Container {
  const config = loadConfig();
  const db = createDatabase({ connectionString: config.databaseUrl, maxConnections: config.dbPoolMax });
  const redis = createRedis(config.redisUrl);
  const repos = createPostgresRepositories();
  const locks = new RedisLockManager(createRedisLockStore(redis));
  const now = () => new Date();

  const reservations = createReservationService({
    db,
    repos,
    locks,
    generateToken: generateReservationToken,
    now,
    reservationTimeoutMinutes: config.reservationTimeoutMinutes,
  });
  const payments = createPaymentService({
    db,
    repos,
    reservations,
    locks,
    gateway: createSimulatedGateway(config.paymentGatewayDelayMs),
    now,
    generateId: randomUUID,
  });
  const catalog = createCatalogService({ db, repos });

  return { config, db, redis, catalog, reservations, payments };
}

const holder = globalThis as unknown as { __ticketingContainer?: Container };

export function getContainer(): Container {
  holder.__ticketingContainer ??= build();
  return holder.__ticketingContainer;
}

File: backend/app/container.py
"""Composição das dependências — equivalente a container.ts. Criado uma única vez por processo,
na inicialização (lifespan) da aplicação FastAPI, e guardado em `app.state`."""
import uuid
from dataclasses import dataclass

from app.config import Settings, get_settings
from app.database import get_session_factory
from app.jobs.expiration import ExpirationJob
from app.lock import RedisLockManager
from app.payment_gateway import SimulatedGateway
from app.redis_client import get_redis
from app.services.catalog_service import CatalogService
from app.services.payment_service import PaymentService
from app.services.reservation_service import ReservationService


@dataclass
class Container:
    config: Settings
    locks: RedisLockManager
    catalog: CatalogService
    reservations: ReservationService
    payments: PaymentService
    expiration_job: ExpirationJob | None = None


def build_container() -> Container:
    config = get_settings()
    session_factory = get_session_factory()
    redis = get_redis()
    locks = RedisLockManager(redis)

    reservations = ReservationService(
        session_factory=session_factory,
        locks=locks,
        reservation_timeout_minutes=config.reservation_timeout_minutes,
    )
    payments = PaymentService(
        session_factory=session_factory,
        reservations=reservations,
        locks=locks,
        gateway=SimulatedGateway(config.payment_gateway_delay_ms),
        generate_id=lambda: str(uuid.uuid4()),
    )
    catalog = CatalogService(session_factory=session_factory)

    return Container(config=config, locks=locks, catalog=catalog, reservations=reservations, payments=payments)
